bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntries.java

```diff
@@ -35,10 +35,15 @@
     LedgerEntry getEntry(long entryId);
 
     /**
-     * In this, It does not retain the ByteBuf references for the entries in this LedgerEntries.
+     * In this method, It does not increment the reference counts of ByteBuf for the entries in this LedgerEntries.
      * The caller who calls {@link #iterator()} should be careful for not releasing the references.
      *
-     * @return the iterator of type LedgerEntry
+     * when iterator is called, you are handing out the entries, you may not know when the caller will
```

Comment:
I don't think this is correct. You're not handing out the entries. You are handing out an iterator which will iterate over the set of entries held by the LedgerEntries. I don't think refcounting should come into the iterator part of it at all.

Reply:
@ivankelly If you think this as `ByteBuf.slice()`, you are handing out a slice of the bytebuffer, you need to declare what would be the behavior for referencing count. Personally I feel having `iterator` and `retainIterator` will make a clear behavior what callers would expect from the interfaces. This is not a `correctness` debate here, it is more about what should be the reference counting semantic for the callers.

Reply:
I didn't mean "correct" in the same way we usually talk about "correctness" :).

Rather, I was arguing semantics. I think the reference count stays with the LedgerEntries object.

Reply:
sure. "reference count" should goes with how caller uses it, not simply stays with LedgerEntries object. that is the semantic problem, you need to take account on how applications use it.

Reply:
The reference count that a LedgerEntry has before iterator() is called belongs to LedgerEntries, because that is referring to the fact that LedgerEntries own those objects.

With retainedIterator there is two options. a) the ownership of the reference is moved to the Iterator (like move semantics in c++) or b) a new Iterator object with a new reference count on the LedgerEntry is created.

a) is a move, b) is a copy (albeit shallow).

If retainedIterator is part of the API I would prefer the a) semantic, but this would imply that retainedIterator can only be called once. If it's b) I would prefer it be done with a Utility method. 

And if retainedIterator is kept, rename it to retainingIterator. It should be an adjective.

Reply:
@ivankelly it is not a `move`. `retain` is to increment the reference count. there is no move here, as the caller don't really know how many reference counts are referencing the underneath buffer. 

The whole idea behind `retainedIterator` is following what `ByteBuf` has and provide exactly same behavior as what people would expect from `ByteBuf`. There is already some sort of reference counting mechanisms that bookkeeper is using. We can just follow what it already has and have a consistent naming, semantic and behavior as `ByteBuf`. Consistent naming and semantic is more important to me.

retain - https://netty.io/4.1/api/io/netty/buffer/ByteBuf.html#retain--
retainedSlice - https://netty.io/4.1/api/io/netty/buffer/ByteBuf.html#retainedSlice-int-int-

Anyway, let's move forward without `retainedIterator`. It is not a blocker for 4.6.0. We can circle back in future if it is needed.

Reply:
> Anyway, let's move forward without retainedIterator. It is not a blocker for 4.6.0. We can circle back in future if it is needed.

Let's leave it out for now then. I think with APIs it's better to leave it out if in doubt. It's always easier to add calls to an API than remove them.

Regarding retain vs retained vs retaining, in the case of ByteBuf, it is the bytebuf itself that has the reference count, so it makes sense to "retain" a bytebuf and have a "retained" bytebuf. However, in the case of an iterator, it's not the iterator that is being retained. Rather the iterator is just a container for objects that are retained, so in some sense the iterator is retaining the objects, though that's not perfect either.

Another thing I thought about with the retainedIterator is that error handling becomes awkward. If while iterating an error occurs, you still need to iterate to the end to free all refcounts, which isn't immediately obvious from the api.

In summary, there seems to be a good few sharp corners on this call, so best to leave it out for now.

Reply:
Thanks @ivankelly and @sijie for your comments. 
I will remove it now, and think of it later when it is wanted.